### ISSUE TITLE
Use `.server.js` and `.client.js` to assert the JavaScript environment

### ISSUE THREAD

**Issue by brillout** [brillout/vite-plugin-ssr#744]
### Description

Now that the V1 design removes the filename suffixes such as `.page.server.js` and `.page.client.js`, we can use `.server.js` and `.client.js` for another purpose.

I suggest we use these to assert where the file is loaded.

```js
// /some/where/SomeClientOnlyComponent.client.js

export function SomeClientOnlyComponent() {
   // ...
}
```

```js
// /some/where/secrets.server.js

export const databaseConnection = {
  password: '4H34CXA3cMa'
}
```

Enabling the user to ensure that the password won't ever be loaded in the browser.

**Comment by dajinchu** [comment]
Is there a way to do this manually in the meantime? Does checking `typeof window` at the top of a file cover it? Or is it more nuanced. Thanks

**Comment by dajinchu** [comment]
Wrote a simple plugin. It's pretty dumb but gets the job done for us. It just checks for substring ".server" in import statements.

https://gist.github.com/dajinchu/2c03ec791ad135e481dbb96bd3538121

Wrote another version using `resolveId` instead of `transform` but transform gives a nice error message showing where the import happened.

**Comment by brillout** [comment]
> a nice error message showing where the import happened.

Good idea 👌. (And the dependency on `es-module-lexer` should be fine.)

I'll additionally implement the "dumber" version with an internal assertion, increasing the safety of this feature.

**Comment by usk94** [comment]
Hi @brillout. I am going to start working on this issue, and I plan to follow this approach for both the server and the client side, is this correct?

> It just checks for substring ".server" in import statements.

**Comment by brillout** [comment]
Yes, I think this approach makes a lot of sense 👍

**Comment by brillout** [comment]
@igolkotek I don't understand your question but feel free to elaborate in a new GH discussion.

**Comment by igolkotek** [comment]
I'm afraid you just didn't understand me. This is not off-topic.
As you say:

```
// /some/where/secrets.server.js
// The file ends with `.server.js` and is therefore guaranteed to never load on the client-side
```
My question is about `onBeforeRender.js` file, if running both on server and client, but with different logic.
So I need to separate code :

```js
if (import.meta.env.SSR) {
...
} else {
// some dynamic import for a client.
}
```

In other words how could I separate onBeforeRender.server.js from onBeforeRender.client.js ? 
Now it is not possible.

**Comment by igolkotek** [comment]
Sure, but if I try to create `onBeforeRender.client.js` server warns me that it will not be executed.

**Comment by brillout** [comment]
See https://vike.dev/meta. I still don't see how your concern is related to this issue. I'm marking our conversation as off-topic, but feel free to create a new GH discussion.
